src/index/base.cpp

```diff
@@ -75,17 +76,28 @@ void BaseIndexNotifications::blockConnected(const interfaces::BlockInfo& block)
     if (!block.data || m_index.IgnoreBlockConnected(block)) return;
 
     const CBlockIndex* best_block_index = m_index.m_best_block_index.load();
-    if (best_block_index != pindex->pprev && !m_index.Rewind(best_block_index, pindex->pprev)) {
+    if (block.chain_tip && best_block_index != pindex->pprev && !m_index.Rewind(best_block_index, pindex->pprev)) {
         FatalError("%s: Failed to rewind index %s to a previous chain tip",
                    __func__, m_index.GetName());
         return;
     }
 
-    if (!m_index.CustomAppend(block)) {
+    interfaces::BlockInfo block_info = node::MakeBlockInfo(pindex, block.data);
+    CBlockUndo block_undo;
+    if (m_options.connect_undo_data && pindex->nHeight > 0) {
+        if (!node::UndoReadFromDisk(block_undo, pindex)) {
+            FatalError("%s: Failed to read block %s undo data from disk",
+                       __func__, pindex->GetBlockHash().ToString());
+            return m_index.Interrupt();
+        }
+        block_info.undo_data = &block_undo;
+    }
+    if (!m_index.CustomAppend(block_info)) {
         FatalError("%s: Failed to write block %s to index",
                    __func__, pindex->GetBlockHash().ToString());
-        return;
-    } else {
+        return m_index.Interrupt();
+    }
+    if (m_index.m_synced) {
```

Comment:
b8be01a7ef20672fd5faf4f6b82cf5d3fa2a717b : this also seems new. Maybe belongs in the next commit?

Reply:
Ok. Current behavior in master (which should not be changed here) still does not make sense to me. I don't think it's wrong (though it could be wrong), I just mean it seems inconsistent and strange.